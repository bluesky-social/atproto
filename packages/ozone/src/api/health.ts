import express from 'express'
import { sql } from 'kysely'
import AppContext from '../context'

export const createRouter = (ctx: AppContext): express.Router => {
  const router = express.Router()

  router.get('/robots.txt', function (req, res) {
    res.type('text/plain')
    res.send(
      '# Hello Friends!\n\n# Crawling the public parts of the API is allowed. HTTP 429 ("backoff") status codes are used for rate-limiting. Up to a handful concurrent requests should be ok.\nUser-agent: *\nAllow: /',
    )
  })

  router.get('/xrpc/_health', async function (req, res) {
    const { version } = ctx.cfg.service
    try {
      await sql`select 1`.execute(ctx.db.db)
    } catch (err) {
      req.log.error(err, 'failed health check')
      return res.status(503).send({ version, error: 'Service Unavailable' })
    }
    res.send({ version })
  })

  return router
}
