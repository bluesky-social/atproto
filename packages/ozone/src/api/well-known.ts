import express from 'express'
import AppContext from '../context'

export const createRouter = (ctx: AppContext): express.Router => {
  const router = express.Router()

  router.get('/.well-known/did.json', (_req, res) => {
    const hostname =
      ctx.cfg.service.publicUrl && new URL(ctx.cfg.service.publicUrl).hostname
    if (!hostname || ctx.cfg.service.did !== `did:web:${hostname}`) {
      return res.sendStatus(404)
    }
    res.json({
      '@context': ['https://www.w3.org/ns/did/v1'],
      id: ctx.cfg.service.did,
      verificationMethod: [
        {
          id: `${ctx.cfg.service.did}#atproto_label`,
          type: 'Multikey',
          controller: ctx.cfg.service.did,
          publicKeyMultibase: ctx.signingKey.did().replace('did:key:', ''),
        },
      ],
      service: [
        {
          id: '#atproto_labeler',
          type: 'AtprotoLabeler',
          serviceEndpoint: `https://${hostname}`,
        },
      ],
    })
  })

  return router
}
