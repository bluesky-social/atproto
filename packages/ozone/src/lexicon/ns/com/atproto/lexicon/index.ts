/**
 * GENERATED CODE - DO NOT MODIFY
 */
import {
  type Auth,
  type Options as XrpcOptions,
  Server as XrpcServer,
  type StreamConfigOrHandler,
  type MethodConfigOrHandler,
  createServer as createXrpcServer,
} from '@atproto/xrpc-server'
import * as ComAtprotoLexiconSchema from '../../../../types/com/atproto/lexicon/schema.js'
import { Server } from '../../../../index.js'

export class ComAtprotoLexiconNS {
  _server: Server

  constructor(server: Server) {
    this._server = server
  }
}
