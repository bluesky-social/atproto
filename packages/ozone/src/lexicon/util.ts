/**
 * GENERATED CODE - DO NOT MODIFY
 */
export function isObj(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null
}

export function hasProp<K extends PropertyKey>(
  data: object,
  prop: K,
): data is Record<K, unknown> {
  return prop in data
}
