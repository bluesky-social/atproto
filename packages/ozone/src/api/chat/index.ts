import { Server } from '../../lexicon'
import AppContext from '../../context'
import getActorMetadata from './getActorMetadata'
import getMessageContext from './getMessageContext'

export default function (server: Server, ctx: AppContext) {
  getActorMetadata(server, ctx)
  getMessageContext(server, ctx)
  return server
}
