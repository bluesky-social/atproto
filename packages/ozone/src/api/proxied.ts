import { Server } from '../lexicon'
import AppContext from '../context'
import { ids } from '../lexicon/lexicons'

export default function (server: Server, ctx: AppContext) {
  server.app.bsky.actor.getProfile({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.actor.getProfile(
        request.params,
        await ctx.appviewAuth(ids.AppBskyActorGetProfile),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.actor.getProfiles({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.actor.getProfiles(
        request.params,
        await ctx.appviewAuth(ids.AppBskyActorGetProfiles),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.getAuthorFeed({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.getAuthorFeed(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedGetAuthorFeed),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.searchPosts({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.searchPosts(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedSearchPosts),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.getPostThread({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.getPostThread(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedGetPostThread),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.getFeedGenerator({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.getFeedGenerator(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedGetFeedGenerator),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getFollows({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getFollows(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetFollows),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getFollowers({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getFollowers(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetFollowers),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getList({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getList(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetList),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getLists({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getLists(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetLists),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.com.atproto.admin.searchAccounts({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      if (!ctx.pdsAgent) {
        throw new Error('PDS not configured')
      }
      const res = await ctx.pdsAgent.api.com.atproto.admin.searchAccounts(
        request.params,
        await ctx.pdsAuth(ids.ComAtprotoAdminSearchAccounts),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getStarterPack({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getStarterPack(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetStarterPack),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getStarterPacks({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getStarterPacks(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetStarterPacks),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getActorStarterPacks({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res =
        await ctx.appviewAgent.api.app.bsky.graph.getActorStarterPacks(
          request.params,
          await ctx.appviewAuth(ids.AppBskyGraphGetActorStarterPacks),
        )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.getLikes({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.getLikes(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedGetLikes),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.getRepostedBy({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.getRepostedBy(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedGetRepostedBy),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.actor.searchActorsTypeahead({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res =
        await ctx.appviewAgent.api.app.bsky.actor.searchActorsTypeahead(
          request.params,
          await ctx.appviewAuth(ids.AppBskyActorSearchActorsTypeahead),
        )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })
}
