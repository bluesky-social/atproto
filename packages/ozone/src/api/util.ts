import { InvalidRequestError } from '@atproto/xrpc-server'
import { InputSchema as ReportInput } from '../lexicon/types/com/atproto/moderation/createReport'
import {
  REASONOTHER,
  REASONSPAM,
  REASONMISLEADING,
  REASONRUDE,
  REASONSEXUAL,
  REASONVIOLATION,
  REASONAPPEAL,
} from '../lexicon/types/com/atproto/moderation/defs'
import { AccountView } from '../lexicon/types/com/atproto/admin/defs'
import {
  RepoView,
  RepoViewDetail,
  REVIEWCLOSED,
  REVIEWESCALATED,
  REVIEWOPEN,
} from '../lexicon/types/tools/ozone/moderation/defs'
import { ModerationEvent } from '../db/schema/moderation_event'
import { ModerationSubjectStatusRow } from '../mod-service/types'
import AppContext from '../context'
import { Member } from '../db/schema/member'
import {
  ROLEADMIN,
  ROLEMODERATOR,
  ROLETRIAGE,
} from '../lexicon/types/tools/ozone/team/defs'
import { ids } from '../lexicon/lexicons'

export const getPdsAccountInfo = async (
  ctx: AppContext,
  did: string,
): Promise<AccountView | null> => {
  const agent = ctx.pdsAgent
  if (!agent) return null
  const auth = await ctx.pdsAuth(ids.ComAtprotoAdminGetAccountInfo)
  if (!auth) return null
  try {
    const res = await agent.api.com.atproto.admin.getAccountInfo({ did }, auth)
    return res.data
  } catch {
    return null
  }
}

export const addAccountInfoToRepoViewDetail = (
  repoView: RepoViewDetail,
  accountInfo: AccountView | null,
  includeEmail = false,
): RepoViewDetail => {
  if (!accountInfo) return repoView
  return {
    ...repoView,
    email: includeEmail ? accountInfo.email : undefined,
    invitedBy: accountInfo.invitedBy,
    invitesDisabled: accountInfo.invitesDisabled,
    inviteNote: accountInfo.inviteNote,
    invites: accountInfo.invites,
    emailConfirmedAt: accountInfo.emailConfirmedAt,
    deactivatedAt: accountInfo.deactivatedAt,
  }
}

export const addAccountInfoToRepoView = (
  repoView: RepoView,
  accountInfo: AccountView | null,
  includeEmail = false,
): RepoView => {
  if (!accountInfo) return repoView
  return {
    ...repoView,
    email: includeEmail ? accountInfo.email : undefined,
    invitedBy: accountInfo.invitedBy,
    invitesDisabled: accountInfo.invitesDisabled,
    inviteNote: accountInfo.inviteNote,
    deactivatedAt: accountInfo.deactivatedAt,
  }
}

export const getReasonType = (reasonType: ReportInput['reasonType']) => {
  if (reasonTypes.has(reasonType)) {
    return reasonType as NonNullable<ModerationEvent['meta']>['reportType']
  }
  throw new InvalidRequestError('Invalid reason type')
}

export const getEventType = (type: string) => {
  if (eventTypes.has(type)) {
    return type as ModerationEvent['action']
  }
  throw new InvalidRequestError('Invalid event type')
}

export const getReviewState = (reviewState?: string) => {
  if (!reviewState) return undefined
  if (reviewStates.has(reviewState)) {
    return reviewState as ModerationSubjectStatusRow['reviewState']
  }
  throw new InvalidRequestError('Invalid review state')
}

const reviewStates = new Set([REVIEWCLOSED, REVIEWESCALATED, REVIEWOPEN])

const reasonTypes = new Set([
  REASONOTHER,
  REASONSPAM,
  REASONMISLEADING,
  REASONRUDE,
  REASONSEXUAL,
  REASONVIOLATION,
  REASONAPPEAL,
])

const eventTypes = new Set([
  'tools.ozone.moderation.defs#modEventTakedown',
  'tools.ozone.moderation.defs#modEventAcknowledge',
  'tools.ozone.moderation.defs#modEventEscalate',
  'tools.ozone.moderation.defs#modEventComment',
  'tools.ozone.moderation.defs#modEventLabel',
  'tools.ozone.moderation.defs#modEventReport',
  'tools.ozone.moderation.defs#modEventMute',
  'tools.ozone.moderation.defs#modEventUnmute',
  'tools.ozone.moderation.defs#modEventMuteReporter',
  'tools.ozone.moderation.defs#modEventUnmuteReporter',
  'tools.ozone.moderation.defs#modEventReverseTakedown',
  'tools.ozone.moderation.defs#modEventEmail',
  'tools.ozone.moderation.defs#modEventResolveAppeal',
  'tools.ozone.moderation.defs#modEventTag',
  'tools.ozone.moderation.defs#modEventDivert',
])

export const getMemberRole = (role: string) => {
  if (memberRoles.has(role)) {
    return role as Member['role']
  }
  throw new InvalidRequestError('Invalid member role')
}

const memberRoles = new Set([ROLEADMIN, ROLEMODERATOR, ROLETRIAGE])
