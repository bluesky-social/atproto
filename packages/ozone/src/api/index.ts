import { Server } from '../lexicon'
import AppContext from '../context'
import createReport from './moderation/createReport'
import emitModerationEvent from './admin/emitModerationEvent'
import searchRepos from './admin/searchRepos'
import adminGetRecord from './admin/getRecord'
import getRepo from './admin/getRepo'
import queryModerationStatuses from './admin/queryModerationStatuses'
import queryModerationEvents from './admin/queryModerationEvents'
import getModerationEvent from './admin/getModerationEvent'
import queryLabels from './label/queryLabels'
import subscribeLabels from './label/subscribeLabels'
import fetchLabels from './temp/fetchLabels'
import createCommunicationTemplate from './admin/createCommunicationTemplate'
import updateCommunicationTemplate from './admin/updateCommunicationTemplate'
import deleteCommunicationTemplate from './admin/deleteCommunicationTemplate'
import listCommunicationTemplates from './admin/listCommunicationTemplates'
import proxied from './proxied'

export * as health from './health'

export * as wellKnown from './well-known'

export default function (server: Server, ctx: AppContext) {
  createReport(server, ctx)
  emitModerationEvent(server, ctx)
  searchRepos(server, ctx)
  adminGetRecord(server, ctx)
  getRepo(server, ctx)
  getModerationEvent(server, ctx)
  queryModerationEvents(server, ctx)
  queryModerationStatuses(server, ctx)
  queryLabels(server, ctx)
  subscribeLabels(server, ctx)
  fetchLabels(server, ctx)
  listCommunicationTemplates(server, ctx)
  createCommunicationTemplate(server, ctx)
  updateCommunicationTemplate(server, ctx)
  deleteCommunicationTemplate(server, ctx)
  proxied(server, ctx)
  return server
}
