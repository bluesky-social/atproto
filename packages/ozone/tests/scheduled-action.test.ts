import { AtpAgent } from '@atproto/api'
import { SeedClient, TestNetwork, basicSeed } from '@atproto/dev-env'
import { ids } from '../src/lexicon/lexicons'
import { forSnapshot } from './_util'

describe('scheduled action management', () => {
  let network: TestNetwork
  let adminAgent: AtpAgent
  let modAgent: AtpAgent
  let triageAgent: AtpAgent
  let sc: SeedClient

  const getAdminHeaders = async (route: string) => {
    return {
      headers: await network.ozone.modHeaders(route, 'admin'),
    }
  }

  const getModHeaders = async (route: string) => {
    return {
      headers: await network.ozone.modHeaders(route, 'moderator'),
    }
  }

  const getModEvent = async (params: {
    subject: string
    cancellation?: boolean
  }) => {
    const {
      data: { events },
    } = await adminAgent.tools.ozone.moderation.queryEvents(
      {
        subject: params.subject,
        types: [
          params.cancellation
            ? 'tools.ozone.moderation.defs#cancelScheduledTakedownEvent'
            : 'tools.ozone.moderation.defs#scheduleTakedownEvent',
        ],
      },
      await getAdminHeaders(ids.ToolsOzoneModerationQueryEvents),
    )
    return events
  }

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'ozone_scheduled_action_test',
    })
    adminAgent = network.ozone.getClient()
    modAgent = network.ozone.getClient()
    triageAgent = network.ozone.getClient()
    sc = network.getSeedClient()
    await basicSeed(sc)
    await network.processAll()
  })

  afterAll(async () => {
    await network.close()
  })

  describe('scheduleAction', () => {
    const getTestAction = () => ({
      action: {
        $type: 'tools.ozone.moderation.scheduleAction#takedown',
        comment: 'test',
        policies: ['spam'],
      },
      subjects: [sc.dids.carol, sc.dids.bob],
      createdBy: 'did:plc:moderator',
      scheduling: {
        executeAt: new Date(Date.now() + 60 * 60 * 1000).toISOString(), // 1 hour from now
      },
    })

    it('allows admins to schedule actions', async () => {
      const { data: result } =
        await adminAgent.tools.ozone.moderation.scheduleAction(
          getTestAction(),
          await getAdminHeaders(ids.ToolsOzoneModerationScheduleAction),
        )
      const bobsModEvents = await getModEvent({ subject: sc.dids.bob })

      expect(result.succeeded.length).toBe(2)
      expect(result.failed.length).toBe(0)
      expect(result.succeeded).toContain(sc.dids.carol)
      expect(result.succeeded).toContain(sc.dids.bob)
      expect(bobsModEvents.length).toBe(1)
    })

    it('rejects triage role from scheduling actions', async () => {
      await expect(
        triageAgent.tools.ozone.moderation.scheduleAction(getTestAction(), {
          headers: await network.ozone.modHeaders(
            ids.ToolsOzoneModerationScheduleAction,
            'triage',
          ),
        }),
      ).rejects.toThrow('Must be a moderator to schedule actions')
    })

    it('supports scheduling with time range (executeAfter/executeUntil)', async () => {
      const rangeAction = {
        ...getTestAction(),
        subjects: [sc.dids.alice],
        scheduling: {
          executeAfter: new Date(Date.now() + 30 * 60 * 1000).toISOString(), // 30 min from now
          executeUntil: new Date(Date.now() + 90 * 60 * 1000).toISOString(), // 90 min from now
        },
      }

      const { data: result } =
        await adminAgent.tools.ozone.moderation.scheduleAction(
          rangeAction,
          await getAdminHeaders(ids.ToolsOzoneModerationScheduleAction),
        )
      expect(result.succeeded.length).toBe(1)
      expect(result.succeeded).toContain(sc.dids.alice)
    })

    it('prevents scheduling multiple actions for same subject', async () => {
      const duplicateAction = {
        ...getTestAction(),
        subjects: [sc.dids.carol],
        scheduling: {
          executeAt: new Date(Date.now() + 3 * 60 * 60 * 1000).toISOString(),
        },
      }

      const { data: result } =
        await adminAgent.tools.ozone.moderation.scheduleAction(
          duplicateAction,
          await getAdminHeaders(ids.ToolsOzoneModerationScheduleAction),
        )
      expect(result.succeeded.length).toBe(0)
      expect(result.failed.length).toBe(1)
      expect(result.failed[0].subject).toBe(sc.dids.carol)
      expect(result.failed[0].error).toContain(
        'A pending scheduled action already exists',
      )
    })

    it('validates scheduling parameters', async () => {
      const invalidAction = {
        ...getTestAction(),
        subjects: ['did:plc:test_invalid'],
        scheduling: {
          executeAfter: new Date(Date.now() + 90 * 60 * 1000).toISOString(),
          executeUntil: new Date(Date.now() + 30 * 60 * 1000).toISOString(), // executeUntil before executeAfter
        },
      }

      const { data } = await adminAgent.tools.ozone.moderation.scheduleAction(
        invalidAction,
        await getAdminHeaders(ids.ToolsOzoneModerationScheduleAction),
      )

      expect(data.failed.length).toBe(1)
      expect(data.failed[0].subject).toBe('did:plc:test_invalid')
      expect(data.failed[0].error).toContain(
        'executeAfter must be before executeUntil',
      )
    })
  })

  describe('listScheduledActions', () => {
    it('allows moderators to list all scheduled actions', async () => {
      const { data: result } =
        await modAgent.tools.ozone.moderation.listScheduledActions(
          {},
          await getModHeaders(ids.ToolsOzoneModerationListScheduledActions),
        )

      expect(result.actions.length).toBeGreaterThan(0)
      expect(forSnapshot(result.actions)).toMatchSnapshot()
    })

    it('allows filtering by subjects', async () => {
      const { data: result } =
        await adminAgent.tools.ozone.moderation.listScheduledActions(
          {
            subjects: [sc.dids.carol],
          },
          await getAdminHeaders(ids.ToolsOzoneModerationListScheduledActions),
        )

      expect(result.actions.length).toBeGreaterThan(0)
      result.actions.forEach((action) => {
        expect(action.did).toBe(sc.dids.carol)
      })
    })

    it('allows filtering by status', async () => {
      const { data: result } =
        await adminAgent.tools.ozone.moderation.listScheduledActions(
          {
            statuses: ['pending'],
          },
          await getAdminHeaders(ids.ToolsOzoneModerationListScheduledActions),
        )

      expect(result.actions.length).toBeGreaterThan(0)
      result.actions.forEach((action) => {
        expect(action.status).toBe('pending')
      })
    })

    it('supports time range filtering', async () => {
      const now = new Date()
      const oneHourAgo = new Date(now.getTime() - 60 * 60 * 1000)
      const twoHoursFromNow = new Date(now.getTime() + 2 * 60 * 60 * 1000)

      const { data: result } =
        await adminAgent.tools.ozone.moderation.listScheduledActions(
          {
            startTime: oneHourAgo.toISOString(),
            endTime: twoHoursFromNow.toISOString(),
          },
          await getAdminHeaders(ids.ToolsOzoneModerationListScheduledActions),
        )

      expect(result.actions.length).toBeGreaterThan(0)
    })

    it('supports pagination', async () => {
      const headers = await getAdminHeaders(
        ids.ToolsOzoneModerationListScheduledActions,
      )
      const { data: page1 } =
        await adminAgent.tools.ozone.moderation.listScheduledActions(
          { limit: 2 },
          headers,
        )

      expect(page1.actions.length).toBe(2)
      expect(page1.cursor).toBeDefined()

      const { data: page2 } =
        await adminAgent.tools.ozone.moderation.listScheduledActions(
          {
            limit: 2,
            cursor: page1.cursor,
          },
          headers,
        )

      expect(page2.actions.length).toBeGreaterThan(0)
      expect(page1.actions.map((a) => a.did)).not.toContain(
        page2.actions[0].did,
      )
    })
  })

  describe('cancelScheduledActions', () => {
    it('allows moderators to cancel scheduled actions', async () => {
      const { data: result } =
        await modAgent.tools.ozone.moderation.cancelScheduledActions(
          {
            subjects: [sc.dids.bob],
          },
          await getModHeaders(ids.ToolsOzoneModerationCancelScheduledActions),
        )
      const bobsModEvents = await getModEvent({
        subject: sc.dids.bob,
        cancellation: true,
      })
      expect(result.succeeded.length).toBe(1)
      expect(result.failed.length).toBe(0)
      expect(result.succeeded).toContain(sc.dids.bob)
      expect(bobsModEvents.length).toBe(1)
    })

    it('allows admins to cancel scheduled actions', async () => {
      const { data: result } =
        await adminAgent.tools.ozone.moderation.cancelScheduledActions(
          {
            subjects: [sc.dids.carol],
          },
          await getAdminHeaders(ids.ToolsOzoneModerationCancelScheduledActions),
        )

      expect(result.succeeded.length).toBe(1)
      expect(result.failed.length).toBe(0)
      expect(result.succeeded).toContain(sc.dids.carol)

      const {
        data: { actions },
      } = await adminAgent.tools.ozone.moderation.listScheduledActions(
        {
          subjects: [sc.dids.carol],
        },
        await getAdminHeaders(ids.ToolsOzoneModerationListScheduledActions),
      )

      expect(actions[0].status).toBe('cancelled')
    })

    it('handles cancellation of non-existent actions', async () => {
      const { data: result } =
        await adminAgent.tools.ozone.moderation.cancelScheduledActions(
          {
            subjects: ['did:plc:nonexistent'],
          },
          await getAdminHeaders(ids.ToolsOzoneModerationCancelScheduledActions),
        )

      expect(result.succeeded.length).toBe(0)
      expect(result.failed.length).toBe(1)
      expect(result.failed[0].did).toBe('did:plc:nonexistent')
      expect(result.failed[0].error).toContain(
        'No pending scheduled actions found',
      )
    })

    it('rejects triage moderators from cancelling actions', async () => {
      await expect(
        triageAgent.tools.ozone.moderation.cancelScheduledActions(
          { subjects: [sc.dids.carol] },
          {
            headers: await network.ozone.modHeaders(
              ids.ToolsOzoneModerationCancelScheduledActions,
              'triage',
            ),
          },
        ),
      ).rejects.toThrow('Must be a moderator to cancel scheduled actions')
    })
  })
})
