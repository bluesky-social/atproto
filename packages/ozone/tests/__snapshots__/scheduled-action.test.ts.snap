// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scheduled action management listScheduledActions allows moderators to list all scheduled actions 1`] = `
Array [
  Object {
    "action": "takedown",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "createdBy": "user(1)",
    "did": "user(0)",
    "eventData": Object {
      "$type": "tools.ozone.moderation.scheduleAction#takedown",
      "comment": "test",
      "policies": Array [
        "spam",
      ],
    },
    "executeAfter": "1970-01-01T00:00:00.000Z",
    "executeUntil": "1970-01-01T00:00:00.000Z",
    "id": 3,
    "randomizeExecution": true,
    "status": "pending",
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
  Object {
    "action": "takedown",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "createdBy": "user(1)",
    "did": "user(2)",
    "eventData": Object {
      "$type": "tools.ozone.moderation.scheduleAction#takedown",
      "comment": "test",
      "policies": Array [
        "spam",
      ],
    },
    "executeAt": "1970-01-01T00:00:00.000Z",
    "id": 2,
    "randomizeExecution": false,
    "status": "pending",
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
  Object {
    "action": "takedown",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "createdBy": "user(1)",
    "did": "user(3)",
    "eventData": Object {
      "$type": "tools.ozone.moderation.scheduleAction#takedown",
      "comment": "test",
      "policies": Array [
        "spam",
      ],
    },
    "executeAt": "1970-01-01T00:00:00.000Z",
    "id": 1,
    "randomizeExecution": false,
    "status": "pending",
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
]
`;
